src/Network/AWS/S3/Cache/Remote.hs

```diff
@@ -415,6 +424,31 @@ runLoggingAWS action onErr onSucc = do
       return def
     Right suc -> onSucc suc
 
+
+-- | Retry the provided action
+retryWith ::
+     ( MonadReader r m
+     , MonadLogger m
+     , HasEnv r
+     , HasMinLogLevel r L.LogLevel
+     , HasNumRetries r Int
+     , MonadIO m
+     )
+  => m (Either a b)
+  -> m (Either a b)
+retryWith action = do
+  conf <- ask
+  eResp <- action
+  case eResp of
+    Left _ -> do
+      if ((conf ^. numRetries) > 0)
+        then do
+          liftIO $ threadDelay (5 * 1000000)  -- microseconds
```

Comment:
5 sec is ok, but I'd rather prefer to have an increasing delay, no need to wait 5 sec if an immediate retry would resolve an issue. Also we need to add some logging, so the user knows why this thing is hanging. It would also be very helpful to know why it is failing, so logging the error would also be useful.

I don't think retrying on all errors make sense, only on `TransportError` would be better. 

@wagdav How about something along those lines:

```haskell

retryWith action = do
  conf <- ask
  let n = conf ^. numRetries
      tshow = T.pack . show
      getErrMsg exc =
        case exc of
          HttpExceptionRequest _ (StatusCodeException resp _) ->
            "StatusCodeException: " <> T.pack (show (responseStatus resp))
          HttpExceptionRequest _ (TooManyRedirects rs) ->
            "TooManyRedirects: " <> T.pack (show (P.length rs))
          HttpExceptionRequest _ (InvalidHeader _) -> "InvalidHeader"
          HttpExceptionRequest _ (InvalidRequestHeader _) -> "InvalidRequestHeader"
          HttpExceptionRequest _ (InvalidProxyEnvironmentVariable name _) ->
            "InvalidProxyEnvironmentVariable: " <> name
          _ -> T.pack (displayException exc)
      go i eResp =
        case eResp of
          Left (TransportError err)
            | i > n -> pure eResp
            | otherwise -> do
              let s = min 9 (i * i)
               -- exponential backoff with at most 9 seconds
              logAWS LevelWarn $ "TransportError - " <> getErrMsg err
              logAWS LevelWarn $
                "Retry " <> tshow i <> "/" <> tshow n <> ". Waiting for " <> tshow s <> " seconds"
              liftIO $ threadDelay (s * 1000000)
              eResp' <- action
              go (i + 1) eResp'
          _ -> pure eResp
  action >>= go 1
```